design-documents/remote-invoker.md

```diff
@@ -0,0 +1,618 @@
+            }
+        })->wait();
+        
+        return $productCreated;
+    }
+    
+    public function delete(string $id): void
+    {
+        //Same stuff: using config, userContext, serializers etc.
+        ....
+    }
+
+....
+
+}
+```
+
+Clearly sending such requests involves a lot of logic and dependencies on other classes,
+and an abstraction (InvokerInterface) which will provide a simple way to pass arguments and remote source data
+to execute a remote service request.
+
+#### Exposing internal logic as an external Web API
+Let's say we have a hypothetical situation: customer places an order _(Checkout)_, but for each item ordered we have to update "ordered" count for related products _(Catalog)_
+```
+Class OrderManagement implements OrderManagementInterface
+{
+ 
+....
+ 
+public function place(OrderInterface $order): void
+{
+ 
+....
+ 
+    foreach ($order->getItems() as $product) {
+        //Calling a service from another domain (Catalog), maybe local, maybe remote
+        $this->productRepository->incrOrdederedCount($product->getId());
+    }
+ 
+....
+ 
+}
+```
+If the remote call to _ProductRepositoryInterface_ service was done via RESTful (or GraphQL) web API the _incrOrdederedCount_ would have
+to be exposed as Web API and could be requested via HTTP by anyone and independently from context which we cannot afford since we can't just increase the counter without an order happening so only another service responsible for orders may call _incrOrdederedCount_.
+ 
+If done via the Invoker the _incrOrdederedCount_ wouldn't be exposed to be requested by HTTP clients via web API and wouldn't show up on
+RESTful endpoints list.
+#### Redundant ACL validations when using existing Web API
+We have a hypothetical situation: admin user with access to orders _(Checkout)_, but not to products _(Catalog)_ changes an order and removes a product from the items list:
+```
+class OrderRepository extends OrderRepositoryInterface
+{
+ 
+....
+ 
+    public function save(OrderInterface $order): void
+    {
+        ....
+        
+        foreach ($removedItems as $product)
+        {
+            //Decreasing ordered count since the product is not ordered anymore, product repository is a Catalog service
+            $this->productRepository->decrOrderedCount($product->getId());
+        }
+    }
+ 
+....
+ 
+}
+```
+If the remote call to _ProductRepositoryInterface::decrOrderedCount_ was done via RESTful web API the exposed endpoint would have
+ACL configured for it so it requires _Catalog edit access_ from admins, but our current admin only has access to _orders_ and the call would fail. However if the Catalog domain modules were not remote and the _ProductRepositoryInterface::decrOrderedCount_ call was made then no ACL validation would occur and we would save the order successsfully.
+ 
+With Invoker such ACL validations wouldn't occur and the remote call would be completed just as if it was local.
```

Comment:
ACL for Web API should be revisited. It's not necessary to introduce another type of APIs (invoker-based) to modify access.
Authentication and authorization can be done on API Gateway level, and in-network calls between services can be done w/o additional validation. Or for better security, there can be additional validation on each service, which will be probably similar to what you propose for the invoker. It depends on how we want to implement it.